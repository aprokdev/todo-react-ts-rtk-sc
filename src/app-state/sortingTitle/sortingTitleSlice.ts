import { createSlice } from '@reduxjs/toolkit';

export type SortingTitleType = string;

export const sortingText = {
    CREATION_DATE: 'CREATION DATE',
    ALPHABET: 'ALPHABET',
    ALPHABET_REVERSE: 'ALPHABET-REVERSE',
};

export const sortingTitleSlice = createSlice({
    name: 'sortingTitle',
    initialState: () =>
        JSON.parse(localStorage.getItem('sortingTitle')) || sortingText.CREATION_DATE,
    reducers: {
        sortByDate: () => {
            return sortingText.CREATION_DATE;
        },
        sortByAlphabet: () => {
            return sortingText.ALPHABET;
        },
        sortByAlphabetReverse: () => {
            return sortingText.ALPHABET_REVERSE;
        },
    },
});

// Action creators are generated for each case reducer function
export const { sortByDate, sortByAlphabet, sortByAlphabetReverse } = sortingTitleSlice.actions;

export default sortingTitleSlice.reducer;
