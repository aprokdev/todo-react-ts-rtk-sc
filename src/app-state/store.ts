import hideCompletedReducer from '@app-state/hideCompleted/hideCompletedSlice';
import isSavedTodosReducer from '@app-state/isSavedTodos/isSavedTodosSlice';
import sortingTitleReducer from '@app-state/sortingTitle/sortingTitleSlice';
import todosReducer from '@app-state/todos/todosSlice';
import { PreloadedState, combineReducers, configureStore } from '@reduxjs/toolkit';

export const rootReducer = combineReducers({
    todos: todosReducer,
    sortingTitle: sortingTitleReducer,
    hideCompleted: hideCompletedReducer,
    isSavedTodos: isSavedTodosReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
    const store = configureStore({
        reducer: rootReducer,
        devTools: true,
        preloadedState,
    });

    let isSavedTodos = true;
    // adds subscribtion on every action:
    store.subscribe(() => {
        // saves current todos and sorting states to LocalStorage only if action type is not "clearLocalStorage":
        isSavedTodos = store.getState().isSavedTodos;
        if (isSavedTodos) {
            localStorage.setItem('listTodos', JSON.stringify(store.getState().todos));
            localStorage.setItem('sortingTitle', JSON.stringify(store.getState().sortingTitle));
        }
        isSavedTodos = true;
    });

    return store;
}

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
