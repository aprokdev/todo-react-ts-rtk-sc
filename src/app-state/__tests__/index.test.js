import { sortingText } from '@app-state/sortingTitle/sortingTitleSlice';
import { setupStore } from '@app-state/store';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { Provider } from 'react-redux';
import { act, create } from 'react-test-renderer';
import { renderWithProvider } from '@src/app-state/test-utils';
import App from '../../app';

const store = setupStore();
function Test() {
    return (
        <Provider store={store}>
            <App />
        </Provider>
    );
}

afterEach(localStorage.clear);

describe('Todo Functionality', () => {
    test('matches snapshot', () => {
        let tree;
        act(() => {
            tree = create(<Test />);
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('checks todos header visiability', async () => {
        const user = userEvent.setup();
        renderWithProvider(<App />);
        const todoInput = screen.getByTestId('todo-input');
        const addTodoBtn = screen.getByTestId('todo-create-btn');
        expect(screen.queryByText(/tasks/i)).toBeNull();
        await user.type(todoInput, 'Test todo');
        await user.click(addTodoBtn);
        expect(screen.queryByText(/tasks/i)).toBeInTheDocument();
    });

    test('creates todo item', async () => {
        const firstTodoText = 'Test todo';
        const secondTodoText = 'Test todo number two!';

        const user = userEvent.setup();
        renderWithProvider(<App />);
        const todoInput = screen.getByTestId('todo-input');
        const addTodoBtn = screen.getByTestId('todo-create-btn');
        expect(screen.queryByText(firstTodoText)).toBeNull();
        await user.type(todoInput, firstTodoText);
        await user.click(addTodoBtn);
        expect(screen.queryByText(firstTodoText)).toBeInTheDocument();
        expect(screen.queryByText(secondTodoText)).toBeNull();
        await user.type(todoInput, secondTodoText);
        await user.click(addTodoBtn);
        expect(screen.queryByText(secondTodoText)).toBeInTheDocument();
    });

    test('check todo works properly', async () => {
        const user = userEvent.setup();
        renderWithProvider(<App />);
        const todoInput = screen.getByTestId('todo-input');
        const addTodoBtn = screen.getByTestId('todo-create-btn');
        await user.type(todoInput, 'Test todo');
        await user.click(addTodoBtn);
        const label = screen.getByTestId('Test todo-label');
        const checkbox = screen.getByTestId('Test todo-cb-input');
        expect(checkbox).not.toBeChecked();
        await user.click(label);
        expect(checkbox).toBeChecked();
        await user.click(screen.getByTestId('Test todo-cb-square'));
        expect(checkbox).not.toBeChecked();
    });

    test('todo is editable', async () => {
        const user = userEvent.setup();
        renderWithProvider(<App />);
        const todoInput = screen.getByTestId('todo-input');
        const addTodoBtn = screen.getByTestId('todo-create-btn');
        await user.type(todoInput, 'Test todo');
        await user.click(addTodoBtn);
        const editBtn = screen.getByText(/edit/i);
        await user.click(editBtn);
        const editField = screen.getByTestId('Test todo-edit-field');
        await user.type(editField, ' edited');
        expect(editField.value).toBe('Test todo edited');
        editField.blur();
    });

    test('deleting todo works properly', async () => {
        const user = userEvent.setup();
        renderWithProvider(<App />);
        const todoInput = screen.getByTestId('todo-input');
        const addTodoBtn = screen.getByTestId('todo-create-btn');
        expect(screen.queryByText('Test todo')).toBeNull();
        await user.type(todoInput, 'Test todo');
        await user.click(addTodoBtn);
        expect(screen.queryByText('Test todo')).toBeInTheDocument();
        const deleteBtn = screen.getByText(/delete/i);
        await user.click(deleteBtn);
        expect(screen.queryByText('Test todo')).toBeNull();
    });

    test('"Hide completed" should hide checked todos', async () => {
        const user = userEvent.setup();
        renderWithProvider(<App />);
        const todoInput = screen.getByTestId('todo-input');
        const addTodoBtn = screen.getByTestId('todo-create-btn');
        // add todo 1
        expect(screen.queryByText('Test todo')).toBeNull();
        await user.type(todoInput, 'Test todo');
        await user.click(addTodoBtn);
        expect(screen.queryByText('Test todo')).toBeInTheDocument();
        // add todo 2
        expect(screen.queryByText('Test todo number two')).toBeNull();
        await user.type(todoInput, 'Test todo number two');
        await user.click(addTodoBtn);
        expect(screen.queryByText('Test todo number two')).toBeInTheDocument();
        // check todo 2
        const label = screen.getByText('Test todo number two');
        const checkbox = screen.getByTestId('Test todo number two-cb-input');
        expect(checkbox).not.toBeChecked();
        await user.click(label);
        expect(checkbox).toBeChecked();

        const filterLabel = screen.queryByText('Hide completed');
        expect(filterLabel).not.toBeChecked();
        // hide checked todos
        await user.click(filterLabel);

        expect(screen.queryByText('Test todo')).toBeInTheDocument();
        expect(screen.queryByText('Test todo number two')).toBeNull();
    });

    test('sorting works properly', async () => {
        const user = userEvent.setup();
        renderWithProvider(<App />);
        const todoInput = screen.getByTestId('todo-input');
        const addTodoBtn = screen.getByTestId('todo-create-btn');
        // add todo 1
        expect(screen.queryByText('Test todo')).toBeNull();
        await user.type(todoInput, 'Test todo');
        await user.click(addTodoBtn);
        expect(screen.queryByText('Test todo')).toBeInTheDocument();
        expect(screen.queryByText(/tasks/i)).toBeInTheDocument();
        // add todo 2
        expect(screen.queryByText('CTest todo number two')).toBeNull();
        await user.type(todoInput, 'CTest todo number two');
        await user.click(addTodoBtn);
        expect(screen.queryByText('CTest todo number two')).toBeInTheDocument();
        // add todo 3
        expect(screen.queryByText('ATest todo number three')).toBeNull();
        await user.type(todoInput, 'ATest todo number three');
        await user.click(addTodoBtn);
        expect(screen.queryByText('ATest todo number three')).toBeInTheDocument();
        // add todo 4
        expect(screen.queryByText('BTest todo number four')).toBeNull();
        await user.type(todoInput, 'BTest todo number four');
        await user.click(addTodoBtn);
        expect(screen.queryByText('BTest todo number four')).toBeInTheDocument();
        // add todo 5
        expect(screen.queryByText('321123')).toBeNull();
        await user.type(todoInput, '321123');
        await user.click(addTodoBtn);
        expect(screen.queryByText('321123')).toBeInTheDocument();

        const header = screen.queryByText(/Sort tasks by: CREATION DATE/i);
        const regExpToGet =
            /Test todo|CTest todo number two|ATest todo number three|BTest todo number four|321123/;

        // after first click on header sorts from A to Z:
        await user.click(header);
        expect(header.innerHTML).toBe('✨ Sort tasks by: ALPHABET');
        // checking saving sorting in LocalStorage:
        expect(JSON.parse(localStorage.getItem('sortingTitle'))).toBe(sortingText.ALPHABET);

        const alphabetSortedExpected = [
            '321123',
            'ATest todo number three',
            'BTest todo number four',
            'CTest todo number two',
            'Test todo',
        ];
        const alphabelSortedLabels = screen
            .queryAllByText(regExpToGet)
            .map((label) => label.innerHTML);
        expect(alphabelSortedLabels).toEqual(alphabetSortedExpected);

        // after second click on header sorts from Z to A:
        await user.click(header);
        expect(header.innerHTML).toBe('✨ Sort tasks by: ALPHABET-REVERSE');
        // checking saving sorting in LocalStorage:
        expect(JSON.parse(localStorage.getItem('sortingTitle'))).toBe(sortingText.ALPHABET_REVERSE);

        const alphabetReverseSortedExpected = [
            'Test todo',
            'CTest todo number two',
            'BTest todo number four',
            'ATest todo number three',
            '321123',
        ];
        const alphabeReverselSortedLabels = screen
            .queryAllByText(regExpToGet)
            .map((label) => label.innerHTML);
        expect(alphabeReverselSortedLabels).toEqual(alphabetReverseSortedExpected);

        // after third click on header sorts by creation time:
        await user.click(header);
        expect(header.innerHTML).toBe('✨ Sort tasks by: CREATION DATE');
        // checking saving sorting in LocalStorage:
        expect(JSON.parse(localStorage.getItem('sortingTitle'))).toBe(sortingText.CREATION_DATE);

        const byDateSortedExpected = [
            'Test todo',
            'CTest todo number two',
            'ATest todo number three',
            'BTest todo number four',
            '321123',
        ];
        const byDateSortedLabels = screen
            .queryAllByText(regExpToGet)
            .map((label) => label.innerHTML);
        expect(byDateSortedLabels).toEqual(byDateSortedExpected);
    });
});

////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////

const localTodos = [
    {
        id: 'asds dsaddbsaddft1660138005899',
        label: 'asds dsaddbsaddft',
        isCompleted: true,
        created: 1660138005899,
    },
    {
        id: 'pouipiuoiuou1660138010767',
        label: 'pouipiuoiuou',
        isCompleted: true,
        created: 1660138010767,
    },
    {
        id: 'werewrewr1660138025187',
        label: 'werewrewr',
        isCompleted: false,
        created: 1660138025187,
    },
    {
        id: 'amfdfd1660138034979',
        label: 'dfgfdamfdfd',
        isCompleted: true,
        created: 1660138034979,
    },
    {
        id: 'iidfigdfigdf1660138040124',
        label: 'iidfigdfigdf',
        isCompleted: true,
        created: 1660138040124,
    },
    {
        id: 'bfsdfdsfds1660138042611',
        label: 'bfsdfdsfds',
        isCompleted: false,
        created: 1660138042611,
    },
    {
        id: '12213fdgd1660140356843',
        label: '12213fdgd',
        isCompleted: false,
        created: 1660140356843,
    },
    {
        id: 'test text1660307378285',
        label: 'test text',
        isCompleted: false,
        created: 1660307378285,
    },
];

const todoListCreationDate = [...localTodos].sort((a, b) => {
    if (a.created < b.created) {
        return -1;
    }
    return 1;
});

const todoListAlphabet = [...localTodos].sort((a, b) => {
    if (a.label < b.label) {
        return -1;
    }
    return 1;
});

const todoListAlphabetReverse = [...localTodos]
    .sort((a, b) => {
        if (a.label < b.label) {
            return -1;
        }
        return 1;
    })
    .reverse();

const orderByCreationDate = [
    'asds dsaddbsaddft',
    'pouipiuoiuou',
    'werewrewr',
    'dfgfdamfdfd',
    'iidfigdfigdf',
    'bfsdfdsfds',
    '12213fdgd',
    'test text',
];
const orderByAlphabet = [
    '12213fdgd',
    'asds dsaddbsaddft',
    'bfsdfdsfds',
    'dfgfdamfdfd',
    'iidfigdfigdf',
    'pouipiuoiuou',
    'test text',
    'werewrewr',
];
const orderByReverseAlphabet = [
    'werewrewr',
    'test text',
    'pouipiuoiuou',
    'iidfigdfigdf',
    'dfgfdamfdfd',
    'bfsdfdsfds',
    'asds dsaddbsaddft',
    '12213fdgd',
];

describe('Todo Functionalityworks with localStorage properly', () => {
    const regExpToGet =
        /asds dsaddbsaddft|pouipiuoiuou|werewrewr|dfgfdamfdfd|iidfigdfigdf|bfsdfdsfds|12213fdgd|test text/;

    test('empty list with clean storage', async () => {
        renderWithProvider(<App />);
        const orderedLabels = screen.queryAllByText(regExpToGet).map((label) => label.innerHTML);
        expect(orderedLabels).toEqual([]);
    });

    test('list is sorted by creation date in case if list exist in storage and sorted by creation date', async () => {
        localStorage.setItem('listTodos', JSON.stringify(todoListCreationDate));
        localStorage.setItem('sortingTitle', JSON.stringify(sortingText.CREATION_DATE));
        renderWithProvider(<App />);
        const orderedLabels = screen.queryAllByText(regExpToGet).map((label) => label.innerHTML);
        expect(orderedLabels).toEqual(orderByCreationDate);
    });

    test('list sorted by alphabet if in storage is list sorted by alphabet', async () => {
        localStorage.setItem('listTodos', JSON.stringify(todoListAlphabet));
        localStorage.setItem('sortingTitle', JSON.stringify(sortingText.ALPHABET));
        renderWithProvider(<App />);
        const orderedLabels = screen.queryAllByText(regExpToGet).map((label) => label.innerHTML);
        expect(orderedLabels).toEqual(orderByAlphabet);
    });

    test('list sorted by alphabet reversed if in storage is list sorted by alphabet reverse', async () => {
        localStorage.setItem('listTodos', JSON.stringify(todoListAlphabetReverse));
        localStorage.setItem('sortingTitle', JSON.stringify(sortingText.ALPHABET_REVERSE));
        renderWithProvider(<App />);
        const orderedLabels = screen.queryAllByText(regExpToGet).map((label) => label.innerHTML);
        expect(orderedLabels).toEqual(orderByReverseAlphabet);
    });
});
