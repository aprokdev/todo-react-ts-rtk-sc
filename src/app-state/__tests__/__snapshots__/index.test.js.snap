// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Functionality matches snapshot 1`] = `
<div
  className="app"
>
  <div
    className="app__head"
  >
    <h1
      className="app__title"
    >
      Todo List
    </h1>
    <img
      alt=""
      className="app__head-img"
      src=""
    />
    <img
      alt=""
      className="app__head-img app__head-img--redux"
      src=""
    />
    <img
      alt=""
      className="app__head-img app__head-img"
      src=""
    />
  </div>
  <form
    className="sc-ewnqHT dkEVeG"
  >
    <input
      className="sc-gTRrQi ftFmhr input"
      data-testid="todo-input"
      id="new-todo-input"
      onChange={[Function]}
      placeholder="Write new task here..."
      type="text"
      value=""
    />
    <button
      className="sc-bgqQcB jbxjhH button"
      data-testid="todo-create-btn"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
